### PR TITLE
Include index issues comments in dts portal

### PR DESCRIPTION
**Pull request by chiaberry**
https://github.com/cityofaustin/atd-data-tech/issues/17974#

The product team needs to be able to easily see the most recent comment on project indices. I added some steps to the gh issues to knack script to get the data the PMs need. 

I added three new columns to `object_30` Project Evaluation in the TEST | 26 APR 2024 | DTS | Data and Technology Services Portal app. 
* Most recent comment, type: paragraph text
* Comment date, type: date/time
* Issue assignees, type: string

I created a new page to display those columns
https://atd.knack.com/test--26-apr-2024--dts--data-and-technology-services-portal#current-projects/

~~The same steps need to be repeated in the prod app, once it is approved.~~
*update! Karo improved my work and added it all to the prod app.* 
You can see it here in prod: https://atd.knack.com/dts#current-projects/ (behind a login)

Screenshot 2024-08-26 at 5 08 24 PM

### To Test

You need the following environment variables from 1pw
ZENHUB_ACCESS_TOKEN
GITHUB_ACCESS_TOKEN
and the knack api key and application id for the DTS Portal, prod

With those variables, run `gh_index_issues_to_dts_portal.py`. Change comments or project assignees data in the Project Evaluations table in the knack builder to force records to update. (Or ask me to do it and I can change it for you)

### CODE REVIEW

**Comment by Charlie-Henry** [comment]
In the issue you linked, it asks for comments to go to the Socrata dataset. Why are we loading them into Knack?

**Comment by chiaberry** [comment]
@Charlie-Henry good question! The socrata dataset has all of the issues, and the knack script only gets the index issues. Amenity said they only care about comments on the index issues, and was comfortable with seeing them in knack. (For now, eventually they want the info on austinmobility.io)

File: gh_index_issues_to_dts_portal.py
Comment (johnclary) [inline gh_index_issues_to_dts_portal.py:1]: just observing that we're a major version behind on `pygithub`: https://github.com/PyGithub/PyGithub/releases 🙃

File: gh_index_issues_to_dts_portal.py [hunk 1]
```diff
@@ -78,30 +93,65 @@ def build_payload(project_records, project_issues):
     for issue in project_issues:  # iterate over gh issues
         pipeline = find_pipeline_by_issue(zenhub_metadata, issue.number)
 
+        last_comment_body = None
+        last_comment_date = None
+        comments = issue.get_comments()
+        # an issue often has more than one assignee, this returns the list of users assigned to the issue
+        assignees = issue.assignees
+        assignees_logins = [user.login for user in assignees]
+        assignees_string = " ".join(assignees_logins)
```
Comment (johnclary) [inline gh_index_issues_to_dts_portal.py:95]: would it be preferable to comma-separate these? or newline (`\n`) them?

File: gh_index_issues_to_dts_portal.py [hunk 2]
```diff
@@ -64,6 +67,18 @@ def find_knack_record_by_issue(knack_records, issue_number):
...
+def are_timestamps_different(knack_timestamp, issue_timestamp):
+    """
+    Returns true if the stored comment timestamp in knack differs from the issues timestamp
+    """
+    if not knack_timestamp:
+        if issue_timestamp:
+            return True
+        else:
+            return False
+    return knack_timestamp["date"] != issue_timestamp.strftime("%m/%d/%Y")
```
Comment (johnclary) [inline gh_index_issues_to_dts_portal.py]: i am trying to think if there's a timezone bug here. i assume `issue_timestamp` is in UTC. and knack doesn't really care what timezone you use. `knackpy` will convert the timestamp to UTC when it loads the data into memory, but i think that only applies to the `knack_timestamp["unix_timestamp"]` property. 

so it looks to me like this will reliably compare apples-to-apples timestamps, but the only edge case concern is that folks may see in Knack that they commented one day ahead of when they actually commented, since here you are saving the calendar date of the timestamp in UTC. i think we would need to test this with a comment left after 7pm CT.  not sure it matters though 🤷

File: gh_index_issues_to_dts_portal.py [hunk 3]
```diff
@@ -78,30 +93,65 @@ def build_payload(project_records, project_issues):
...
+            )
 
             if title_knack != issue.title:
                 issue_payload[KNACK_TITLE_FIELD] = issue.title
+                update_record = True
             if pipeline_knack != pipeline:
                 issue_payload[KNACK_PIPELINE_FIELD] = pipeline
-            if title_knack != issue.title or pipeline_knack != pipeline:
+                update_record = True
+            if are_timestamps_different(last_comment_date_knack, last_comment_date):
```
Comment (mddilley) [inline gh_index_issues_to_dts_portal.py]: I found myself in a situation where I edited a **Most recent comment** in `object_30` but this code was not seeing that the **Comment Date** (`field_689`) was different because I didn't edit that field. Once I edited the date field in the Knack table to be one day in the past, I was able to restore the comment date and most recent comment that is fetched from GitHub.

I'm not sure if I'm thinking about this right, but what would happen if two comments were added on the same day in the GitHub issue? I think the code might not see the update.

File: gh_index_issues_to_dts_portal.py [hunk 4]
```diff
@@ -78,30 +93,65 @@ def build_payload(project_records, project_issues):
     for issue in project_issues:  # iterate over gh issues
         pipeline = find_pipeline_by_issue(zenhub_metadata, issue.number)
 
+        last_comment_body = None
+        last_comment_date = None
+        comments = issue.get_comments()
```
Comment (johnclary) [inline gh_index_issues_to_dts_portal.py:91]: i was wondering if there is some way to avoid having to do this check for every project issue. like if it's closed, can we ignore updates to it? but obviously we would miss updates in that case.

this is slow, but not as slow as i would expect given how much data it's pulling down 👍

**Review by johnclary — APPROVED** [review]
i ran a few tests and all is working just as promised. nice enhancement!!

my feedback is just questions 🤔 

🚢 🚢 🚢 🚢

**Review by mddilley — COMMENTED** [review]
Test steps checked out, and I ran into what might be an edge case so I left a question. Love the new logging! 💯

File: gh_index_issues_to_dts_portal.py (hunk 3 above)
Comment (chiaberry) [reply gh_index_issues_to_dts_portal.py]: hm so the dag runs once a day and I was thinking it would pick up the most recent comment then, since it wouldnt have seen the first comment yet. But there could be an edge case (like how John has seen about timezones too) where it picks up a comment but then the dag runs and doesn't pick up the newest...I think I need to write the times out to see

File: gh_index_issues_to_dts_portal.py (hunk 4 above)
Comment (chiaberry) [reply gh_index_issues_to_dts_portal.py:91]: I'll check with @amenity if its okay to ignore closed projects's comments

File: gh_index_issues_to_dts_portal.py (hunk 3 above)
Comment (johnclary) [reply gh_index_issues_to_dts_portal.py]: perhaps we don't need to compare timestamps, we can just check if the comment is different? that's all that matters, right?

**Comment by johnclary** [comment]
@chiaberry one more question—do we have concerns about markdown and html being included in comments? one fancy option would be to convert markdown to html, in which case i believe you could send it to a knack "paragraph" field and it would render as rich text in the UI

**Comment by chiaberry** [comment]
@johnclary we do have concerns about it, and i didnt think about that idea, i like it

File: gh_index_issues_to_dts_portal.py (hunk 3 above)
Comment (amenity) [reply gh_index_issues_to_dts_portal.py]: Ooh, good catch. This actually isn't an edge case—I regularly polish existing comments on index issues. I like the idea of comparing the comment content instead of the timestamp.

File: gh_index_issues_to_dts_portal.py (hunk 4 above)
Comment (amenity) [reply gh_index_issues_to_dts_portal.py:91]: @chiaberry @johnclary -What are the disadvantages of a slow script?

File: gh_index_issues_to_dts_portal.py (hunk 2 above)
Comment (amenity) [reply gh_index_issues_to_dts_portal.py]: It's unlikely that a day-off date on a comment made after 7pm would be problematic.

FWIW, we're using the date on the front end to filter down to [in-progress projects that haven’t had a status update since the previous newsletter was compiled](https://atd.knack.com/test--26-apr-2024--dts--data-and-technology-services-portal#current-projects/?view_395_filters=%7B%22match%22%3A%22and%22%2C%22rules%22%3A%5B%7B%22field%22%3A%22field_649%22%2C%22operator%22%3A%22contains%22%2C%22value%22%3A%22progress%22%2C%22field_name%22%3A%22Pipeline%22%7D%2C%7B%22match%22%3A%22and%22%2C%22field%22%3A%22field_689%22%2C%22operator%22%3A%22is%20before%22%2C%22value%22%3A%7B%22date%22%3A%2202%2F14%2F2024%22%7D%2C%22field_name%22%3A%22Comment%20date%22%7D%5D%7D&view_395_page=1).

File: gh_index_issues_to_dts_portal.py (hunk 1 above)
Comment (amenity) [reply gh_index_issues_to_dts_portal.py:95]: Nbd, but newline would be nice.

File: gh_index_issues_to_dts_portal.py (hunk 2 above)
Comment (chiaberry) [reply gh_index_issues_to_dts_portal.py]: I am comparing on comment text now

File: gh_index_issues_to_dts_portal.py (hunk 1 above)
Comment (chiaberry) [reply gh_index_issues_to_dts_portal.py:95]: I added a comma. I tried a newline, but the formatting didnt come through in knack, i think its a limitation of the short text field

File: gh_index_issues_to_dts_portal.py [hunk 5]
```diff
@@ -118,10 +163,12 @@ def main():
     repo = g.get_repo(REPO)
 
     # iterate over the github client's issues and build our working data
+    logging.info("Downloading issues from github")
```
Comment (frankhereford) [inline gh_index_issues_to_dts_portal.py:166]: It's the little things -- thanks for adding these `info` log entries. 🙏

**Review by frankhereford — APPROVED** [review]
I've gone over the code carefully, and I think this looks great. I have not run it in the interest of not having to get in knack and change those records or build up the environment -- but I'm more than comfortable with the executions of the software done by previous testers and what is shown in the artifacts in your comment. 

🚢🚢🚢🚢🚢

Thank you @chiaberry !!
